usr/libexec/playtron/hardware-test-tool

```diff
@@ -78,6 +78,59 @@ class InputTest(PlaceholderTest):
         self.done = True
 
 
+class BatteryTest(PlaceholderTest):
+    def __init__(self):
+        self.name = 'Battery|电池'
+        self.reset()
+        self.start_time = None
+        self.battery_path = '/sys/class/power_supply'
+        self.battery_name = None
+        self.battery_percentage = None
+        self.battery_state_text = None
+        self.battery_charging_speed = None
+
+    def start(self, screen, font):
+        self.reset()
+        self.start_time = time.perf_counter()
+
+        for file_name in os.listdir(self.battery_path):
+            if 'BAT' in file_name:
+                self.battery_name = file_name
+        if not self.battery_name:
+            self.result = 'fail'
+            self.done = True
+
+        with open(self.battery_path + '/' + self.battery_name + '/' + 'capacity', 'r') as file_name:
+            self.battery_percentage = file_name.read().strip() + '%'
+
+        # State is either "Charging", "Discharging", or "Not charging".
+        with open(self.battery_path + '/' + self.battery_name + '/' + 'status', 'r') as file_name:
+            battery_state = file_name.read().strip()
+        if battery_state == "Charging":
+            self.battery_state_text = 'Charging|充电'
+        elif battery_state == "Discharging":
+            self.battery_state_text = 'Discharging|放电'
+        else:
+            self.battery_state_text = '?'
+
+        battery_charging_speed_cmd = os.popen(
+            'upower -i /org/freedesktop/UPower/devices/battery_' + self.battery_name + ' | grep energy-rate: | awk \'{print $2}\' | cut -d. -f1')
```

Comment:
This is a bit ugly but it's the only reliable way I could get battery charging speed on the AYANEO 2S. Other handheld devices (ASUS ROG Ally, Steam Deck LCD, etc.) exposed the necessary `/sys` interfaces that allowed me to manually calculate the speed. Surely that information must be exposed somewhere because `upower` is able to get it.